pkg/scheduler/framework/plugins/noderesources/fit.go

```diff
@@ -101,13 +101,150 @@ func getPodResourceRequest(cycleState *framework.CycleState) (*nodeinfo.Resource
+	for _, r := range insufficientResources {
+		failureReasons = append(failureReasons, getErrReason(r.ResourceName))
+	}
+
+	if len(insufficientResources) != 0 {
+		return framework.NewStatus(framework.Unschedulable, failureReasons...)
+	}
+	return nil
+}
+
+func getErrReason(rn v1.ResourceName) string {
+	return fmt.Sprintf("Insufficient %v", rn)
+}
+
+// getResourceRequest returns a *schedulernodeinfo.Resource that covers the largest
+// width in each resource dimension. Because init-containers run sequentially, we collect
+// the max in each dimension iteratively. In contrast, we sum the resource vectors for
+// regular containers since they run simultaneously.
+//
+// If Pod Overhead is specified and the feature gate is set, the resources defined for Overhead
+// are added to the calculated Resource request sum
+//
+// Example:
+//
+// Pod:
+//   InitContainers
+//     IC1:
+//       CPU: 2
+//       Memory: 1G
+//     IC2:
+//       CPU: 2
+//       Memory: 3G
+//   Containers
+//     C1:
+//       CPU: 2
+//       Memory: 1G
+//     C2:
+//       CPU: 1
+//       Memory: 1G
+//
+// Result: CPU: 3, Memory: 3G
+func getResourceRequest(pod *v1.Pod) *schedulernodeinfo.Resource {
+	result := &schedulernodeinfo.Resource{}
+	for _, container := range pod.Spec.Containers {
+		result.Add(container.Resources.Requests)
+	}
+
+	// take max_resource(sum_pod, any_init_container)
+	for _, container := range pod.Spec.InitContainers {
+		result.SetMaxResource(container.Resources.Requests)
+	}
+
+	// If Overhead is being utilized, add to the total requests for the pod
+	if pod.Spec.Overhead != nil && utilfeature.DefaultFeatureGate.Enabled(features.PodOverhead) {
+		result.Add(pod.Spec.Overhead)
+	}
+
+	return result
+}
+
+// InsufficientResource describes what kind of resource limit is hit and caused the pod to
+// not fit the node.
+type InsufficientResource struct {
+	ResourceName v1.ResourceName
+	Requested    int64
+	Used         int64
+	Capacity     int64
+}
+
+// NewInsufficientResourceError returns an InsufficientResourceError.
+func newInsufficientResource(resourceName v1.ResourceName, requested, used, capacity int64) *InsufficientResource {
+	return &InsufficientResource{
+		ResourceName: resourceName,
+		Requested:    requested,
+		Used:         used,
+		Capacity:     capacity,
+	}
+}
+
+// PodFitsResourcesPredicate checks if a node has sufficient resources, such as cpu, memory, gpu, opaque int resources etc to run a pod.
+// First return value indicates whether a node has sufficient resources to run a pod while the second return value indicates the
+// predicate failure reasons if the node has insufficient resources to run the pod
+func PodFitsResourcesPredicate(
+	pod *v1.Pod,
+	podRequest *schedulernodeinfo.Resource,
+	ignoredExtendedResources sets.String,
+	nodeInfo *schedulernodeinfo.NodeInfo,
+) []*InsufficientResource {
+	var insufficientResources []*InsufficientResource
```

Comment:
I know it's pre-existing, but we don't need put pointers into a slice as slice is underneath holding pointers.

Reply:
done.